File: engine/tests/test_minimax.cpp
/**
 * @file test_minimax.cpp
 * @brief Unit tests for the minimax search algorithm.
 */

#include <catch2/catch_test_macros.hpp>
#include <kribu/board.hpp>
#include <kribu/minimax.hpp>
#include <kribu/rules.hpp>

using namespace kribu::board;
using namespace kribu::sholoGuti;

TEST_CASE("Minimax Initial State Evaluation", "[minimax]") {  // NOLINT(readability-function-cognitive-complexity)
  boardState state = INITIAL_STATE;
  REQUIRE(evaluate_board(state) == 0);
}

TEST_CASE("Minimax Terminal State Evaluation", "[minimax]") {  // NOLINT(readability-function-cognitive-complexity)
  boardState winState;
  winState.me = 1ULL << 0;
  winState.opp = 0;
  MinimaxResult resWin = minimax(winState, 2, -INFINITY_VAL, INFINITY_VAL);
  REQUIRE(resWin.score >= INFINITY_VAL);
  REQUIRE(resWin.moveId == -1);

  boardState loseState;
  loseState.me = 0;
  loseState.opp = 1ULL << 0;
  MinimaxResult resLose = minimax(loseState, 2, -INFINITY_VAL, INFINITY_VAL);
  REQUIRE(resLose.score <= -INFINITY_VAL);
  REQUIRE(resLose.moveId == -1);
}

TEST_CASE("Minimax Choose Winning Move", "[minimax]") {  // NOLINT(readability-function-cognitive-complexity)
  // Setup a single capture that wins the game
  boardState state;
  state.me = 1ULL << 16;
  state.opp = 1ULL << 17;
  state.activeCaptureIdx = -1;

  // The move 16 -> 18 captures 17
  int winMoveId = find_move(16, 18);
  REQUIRE(winMoveId != -1);
  REQUIRE(is_capture_move(winMoveId));

  MinimaxResult res = minimax(state, 1, -INFINITY_VAL, INFINITY_VAL);
  REQUIRE(res.moveId == winMoveId);
  REQUIRE(res.score >= INFINITY_VAL);
}

File: python/tests/test_minimax.py
from kribu import (
    INITIAL_STATE,
    minimax,
    boardState,
    find_move,
)


def test_minimax_initial_state():
    # Run minimax at depth 1
    res = minimax(INITIAL_STATE, 1)
    assert hasattr(res, "score")
    assert hasattr(res, "moveId")
    assert isinstance(res.score, int)
    assert isinstance(res.moveId, int)
    # The move ID should be a valid move (greater than 0 since INITIAL_STATE has only simple moves)
    assert res.moveId > 0


def test_minimax_winning_move():
    # Setup winning capture move
    state = boardState()
    state.me = 1 << 16
    state.opp = 1 << 17
    state.activeCaptureIdx = -1

    res = minimax(state, 1)
    win_move_id = find_move(16, 18)
    assert res.moveId == win_move_id
    assert res.score >= 1000000
